### DISCUSSION TITLE
Any component inside <NextIntlClientProvider> will always be re-rendered?

### DISCUSSION THREAD

**Discussion by janvorwerk** [amannn/next-intl#617] [Q&A] [answered]
Hello,

I am using 3.0.0-rc.7 with a mix of server components and client components so I dup'ed the setup, using the `i18n.ts` and the `<NextIntlClientProvider>` for all components to render properly.

Additionaly, on my root layout I have a few client components that must be rendered **exactly once** on the client side (to install global message listeners for instance).

So far so good. The surprising fact is that those components will render upon **each** client-side navigation if I put them inside the `<NextIntlClientProvider>` (and only once if I put them just outside).

Is this expected? How comes? I thought about removing the `now` parameter, which I suppose may cause this:

      now={new Date()}

Are you guys observing the same effect?

Thanks for reading so far 🙏

**Comment by amannn** [comment, marked as answer]
I've addressed this in https://github.com/amannn/next-intl/pull/149/commits/a0a1f045dd6ce1b347a37f38ee39bb1e1aa8987e, this should help! This will be published as part of the next release.

Note however:
1. `NextIntlClientProvider` can only avoid re-rendering consumers, if it receives stable references as props (e.g. if `timeZone` stays consistent, this isn't a problem, but `formats={{…}}` will create a new object reference on every render. In this case, you should add `useMemo` around `formats` if re-renders are important to your app.
2. You [should always be ready to render](https://overreacted.io/writing-resilient-components), avoiding re-renders should only be about performance.

Hope this helps!

**Reply by janvorwerk** [reply]
Thanks for the fix & for the hint about that article. Indeed, I am a beginner with React, so this will definitly help! 🙏

I guess, there is no real issue (besides performance) in re-rendering my component that registers a global `"message"` listener... (provided, I unregister in the cleanup function returned by useEffect())...  

I am a bit scared though, that I could possibly miss messages from other windows between the `removeEventListener("message", ...)` and its subsequent `addEventListener("message", ...)`. Not sure about the gory details about this (is the useEffect() guaranteed to be called in the same event loop as the cleanup of the previous effect?). That's why I thought I'd try to remain safe and render it once and only once if possible.
